app/domain/authentication/authn_k8s/k8s_host.rb

```diff
@@ -12,22 +13,29 @@
 
 module Authentication
   module AuthnK8s
+
+    Err ||= Errors::Authentication::AuthnK8s
+
     class K8sHost
       extend Forwardable
 
       attr_reader :account, :service_name, :csr
 
       def self.from_csr(account:, service_name:, csr:)
         cn = csr.common_name
-        raise ArgumentError, 'CSR must have a CN entry' unless cn
-
+        unless cn
+          raise Err::CSRMissingCNEntry.new(csr.subject_to_s, csr.spiffe_id.to_s, account, service_name)
+        end
         new(account: account, service_name: service_name, common_name: cn)
       end
 
       def self.from_cert(account:, service_name:, cert:)
-        cn = ::Util::OpenSsl::X509::SmartCert.new(cert).common_name
-        raise ArgumentError, 'Certificate must have a CN entry' unless cn
-
+        smart_cert = ::Util::OpenSsl::X509::SmartCert.new(cert)
+        cn = smart_cert.common_name
+        unless cn
+          raise Err::CertMissingCNEntry.new(
+              smart_cert.smart_subject.to_s, smart_cert.san.to_s, smart_cert.san_uri.to_s, account, service_name)
+        end
```

Comment:
if you decide to make this multiline, you should move also the closing bracket and separate each param:
```
raise Err::CertMissingCNEntry.new(
   smart_cert.smart_subject.to_s, 
  smart_cert.san.to_s, 
  smart_cert.san_uri.to_s, 
  account, 
  service_name
)
```